Fix #3963. Update setup-google-drive-remote.md (#4077)
iterative/dvc.org: content/docs/user-guide/how-to/setup-google-drive-remote.md

running inside a Compute Engine, automatic CI/CD, etc. No interactive user OAuth
authentication is needed.

> ℹ️ Google service accounts have their own associated usage limits which may be
> exceeded if used frequently in conjunction with `dvc push/pull`. For heavy usage
> it is recommended to rely on
> [delegation](#delegation-with-google-service-accounts).

Reviewer: Just improved formatting (ignore...)
```suggestion
<admon type="info">

Google service accounts have their own associated usage limits which may be
exceeded if used frequently in conjunction with `dvc push/pull`. For heavy usage
it is recommended to rely on
[delegation](#delegation-with-google-service-accounts).

</admon>
```

running inside a Compute Engine, automatic CI/CD, etc. No interactive user OAuth
authentication is needed.

<admon type="info">

Google service accounts have their own associated usage limits which may be

> exceeded if used frequently in conjunction with `dvc push`, dvc pull`, etc. For

Reviewer: Oops
```suggestion
exceeded if used frequently in conjunction with `dvc push`, `dvc pull`, etc. For
```

> exceeded if used frequently in conjunction with `dvc push/pull`. For heavy usage
> it is recommended to rely on

Reviewer: Copy edit (committing...)
```suggestion
exceeded if used frequently in conjunction with `dvc push`, dvc pull`, etc. For
heavy usage, it is recommended to rely on
```

> [delegation](#delegation-with-google-service-accounts).

Reviewer: Thanks @JohnTheDeere . TBH I haven't tried this in a long time so I'm not sure how all the details work, but I'd like to clarify something: In the top of the doc we state,

> DVC uses the Google Drive API ..., so it's subject to certain usage limits and quotas...
For heavy use, it's highly recommended to [connect using a custom Google Cloud project](https://dvc.org/doc/user-guide/how-to/setup-google-drive-remote#using-a-custom-google-cloud-project-recommended)

Then, that section ([Using a custom Google Cloud project](https://dvc.org/doc/user-guide/how-to/setup-google-drive-remote#using-a-custom-google-cloud-project-recommended)) says,

> We highly recommend this for heavy use and advanced needs because: ...
> - [Using a service account](https://dvc.org/doc/user-guide/how-to/setup-google-drive-remote#using-service-accounts) for automation tasks (e.g. CI/CD) is only possible this way.

And now you're proposing to add a note in that last section ([Using service accounts](https://dvc.org/doc/user-guide/how-to/setup-google-drive-remote#using-service-accounts)) about it's own rate limits (which was the original problem leading us here) -- seems to go in circles?

Sorry, this is a complicated topic and question, you don't have to have an answer. 🙂

Author: Hey @jorgeorpinel yes let's not make it too complicated :D Goal should be to simplify as much as possible imo. To be honest when I was starting to use DVC with google and looking at that documentation there were some moments of desperation understanding how to do things "the right way". Google sure makes things complicated very fast.

Anyway - I did follow those instructions and setup a custom google cloud project + service account. I used the service account to connect to my personal google drive. At this point I was hoping everything would be well. And it did work for a while. Until suddenly I was unable to use the service account to pull/push data. The service account itself has quota limits so without adding delegation to it so you may still face problems regarding uploading/downloading. Knowing about delegation seems like a good thing.

Not sure if this clarified anything... Hoping it at least didn't make anything worse.

Reviewer: Yep this doc just reflects Google Drive's integration complexity... It's def. better to have the info, and thanks for sharing your experience!

   account as an editor (read/write) or viewer (read-only):

![](/img/gdrive-share-with-service-account.png)

### Delegation

[Delegation](https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority)
can be used to overcome quota limits associated with Google service accounts.

> The required **OAuth scope** is `https://www.googleapis.com/auth/drive`
> 
> The remote must also be configured with the associated user **personal email**
> using:

Reviewer: ```suggestion
The required **OAuth scope** is `https://www.googleapis.com/auth/drive`.

The remote must also be configured with the associated user **personal email**:
```

   account as an editor (read/write) or viewer (read-only):

![](/img/gdrive-share-with-service-account.png)

> ## Delegation with google service accounts
> 
> Delegation can be used to overcome quota limits associated with google service
> accounts. See
> [these](https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority)
> instructions for details.

Reviewer: More edits
```suggestion
### Delegation

[Delegation](https://developers.google.com/identity/protocols/oauth2/service-account#delegatingauthority)
can be used to overcome quota limits associated with Google service accounts.
```
